Migrate SecTrust to WebKitSecureCoding Interface

sheeparegreat opened the pull request:
#### 57afa59fdb669f391e5c12656a7e77da6a7b1655
<pre>
Migrate SecTrust to WebKitSecureCoding Interface
<a href="https://bugs.webkit.org/show_bug.cgi?id=284795">https://bugs.webkit.org/show_bug.cgi?id=284795</a>
<a href="https://rdar.apple.com/136786582">rdar://136786582</a>

Reviewed by Sihui Liu.

* Source/WTF/wtf/spi/cocoa/SecuritySPI.h:
* Source/WebKit/Shared/cf/CoreIPCSecTrust.h:
(WebKit::CoreIPCSecTrust::CoreIPCSecTrust):
* Source/WebKit/Shared/cf/CoreIPCSecTrust.mm: Added.
(WebKit::arrayElementsTheSameType):
(WebKit::CoreIPCSecTrust::detectPolicyOptionShape):
(WebKit::updatePolicyVector):
(WebKit::optionalArrayOfDataHelper):
(WebKit::CoreIPCSecTrust::CoreIPCSecTrust):
(WebKit::createPolicyDictionary):
(WebKit::addToDictFromOptionalDataHelper):
(WebKit::CoreIPCSecTrust::createSecTrust const):
* Source/WebKit/Shared/cf/CoreIPCSecTrust.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(secTrustRefsEqual):
(TEST(IPCSerialization, SecTrustRef)):

Canonical link: <a href="https://commits.webkit.org/289561@main">https://commits.webkit.org/289561@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a8a8bed1f605b61591f9e72551d808313c8b350f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87148 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6658 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41507 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92009 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37887 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89198 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6934 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14726 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67351 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25095 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90150 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5309 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78879 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47671 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5082 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33290 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37004 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/80154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75571 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34134 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93894 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/85922 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14310 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10406 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76155 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14514 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74735 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75408 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18574 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19698 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18136 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7227 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14329 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19622 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/108414 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14074 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26085 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17517 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15855 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sheeparegreat commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,343 @@
+            }
+            NSDate *value = [info objectForKey:key];
+            if (![value isKindOfClass:NSDate.class]) {
+                ASSERT_NOT_REACHED();
```
Hit this ASSERT in testing

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,343 @@
+    if (!trust)
+        return;
+
+    auto dict = adoptNS(bridge_cast(static_cast<CFDictionaryRef>(SecTrustCopyPropertyListRepresentation(trust, &error))));
```
We should use dynamic cast:

```
RetainPtr cfDictionary = dynamic_cf_cast<CFDictionaryRef>(adoptCF(SecTrustCopyPropertyListRepresentation(trust, &error)));
if (!dictionary)
    return;

RetainPtr dictionary = bridge_cast(cfDictionary);
```

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,343 @@
+
+    CoreIPCSecTrustData secTrustData { };
+
+    NSDate *verifyDate = [dict objectForKey:@"verifyDate"];
```
`RetainPtr`

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,343 @@
+    }
+ALLOW_DEPRECATED_DECLARATIONS_END
+
+    NSNumber *anchorsOnly = [dict objectForKey:@"anchorsOnly"];
```
`RetainPtr anchorsOnly`

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,343 @@
+    }
+
+ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    NSNumber *result = [dict objectForKey:@"result"];
```
`RetainPtr` as we prefer smart pointer everywhere nowadays

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,343 @@
+            break;
+        case kSecTrustResultOtherError:
+            secTrustData.result = CoreIPCSecTrustResult::OtherError;
+            break;
```
Should we return early or assert if the result does not match any valid value?

sheeparegreat replied:
If we are serializing a SecTrust representing a failed trust evaluation, I think this is what we want to do

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,343 @@
+                return;
+            }
+            auto c = CoreIPCData(item);
+            vector.append(WTFMove(c));
```
Can we do `vector.append(CoreIPCData { item });`

sheeparegreat replied:
For all of these you get a compile error saying that it cannot do copy elision

szewai reviewed (commented):
Haven't done a full round of review yet and will continue after you fix the assertion; a note is that we should be using `RetainPtr` whenever possible

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+{
+    if (!array.count)
+        return true;
+    Class firstElement = [array[0] class];
```
You could use `firstObject`.
Also it seems the callers already know the class before invoking this, so maybe this function should directly take Class.

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+    CoreIPCSecTrustData secTrustData { };
+
+    RetainPtr<NSDate> verifyDate = [dict objectForKey:@"verifyDate"];
+    if ([verifyDate isKindOfClass:NSDate.class]) {
```
Early return to avoid more work if `verifyDate` is not `NSDate`

sheeparegreat replied:
verifyDate is optional in the dictionary so it might not be present

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+
+ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    RetainPtr<NSNumber> result = [dict objectForKey:@"result"];
+    if ([result isKindOfClass:NSNumber.class]) {
```
Early return to avoid more work if `result` is not `NSNumber`

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    RetainPtr<NSNumber> result = [dict objectForKey:@"result"];
+    if ([result isKindOfClass:NSNumber.class]) {
+        switch ([result unsignedCharValue]) {
```
We might have a helper function `CoreIPCSecTrustResult toCoreIPCSecTrustResult(uint8_t)` so here we can just:
`secTrustData.result = toCoreIPCSecTrustResult([result unsignedCharValue])`

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+ALLOW_DEPRECATED_DECLARATIONS_END
+
+    RetainPtr<NSNumber> anchorsOnly = [dict objectForKey:@"anchorsOnly"];
+    if ([anchorsOnly isKindOfClass:NSNumber.class])
```
Early return to avoid more work if `anchorsOnly` is not `NSNumber`

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+        secTrustData.anchorsOnly = [anchorsOnly boolValue];
+
+    RetainPtr<NSArray> certificates = [dict objectForKey:@"certificates"];
+    if ([certificates isKindOfClass:NSArray.class]) {
```
Early return to avoid more work if `certificates` is not `NSArray`

sheeparegreat replied:
This currently checks for nil and the correct type.  If we do get an empty array here, I don't think this is a problem.

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+    }
+
+    RetainPtr<NSArray> chain = [dict objectForKey:@"chain"];
+    if ([chain isKindOfClass:NSArray.class]) {
```
Early return to avoid more work if `chain` is not `NSArray`

sheeparegreat replied:
I think this one can be nil/empty as there might not be a certificate chain.  I think this is ok.

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+    }
+
+    RetainPtr<NSArray> details = [dict objectForKey:@"details"];
+    if ([details isKindOfClass:NSArray.class]) {
```
Early return if `details` is not `NSArray`

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+    }
+
+    RetainPtr<NSDictionary> info = [dict objectForKey:@"info"];
+    if ([info isKindOfClass:NSDictionary.class]) {
```
Early return if `info` is not `NSDictionary`

sheeparegreat replied:
Info is optional so I think this is ok

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+                ASSERT_NOT_REACHED();
+                return;
+            }
+            auto k = CoreIPCString(key);
```
`CoreIPCString keyString { key };`

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+    }
+
+    RetainPtr<NSNumber> keychainsAllowed = [dict objectForKey:@"keychainsAllowed"];
+    if ([keychainsAllowed isKindOfClass:NSNumber.class])
```
Early return if `keychainsAllowed` is not `NSNumber`

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+        secTrustData.keychainsAllowed = [keychainsAllowed boolValue];
+
+    RetainPtr<NSArray> policies = [dict objectForKey:@"policies"];
+    if ([policies isKindOfClass:NSArray.class]) {
```
Return early if `policies` is not `NSArray`

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+    m_data = WTFMove(secTrustData);
+}
+
+static void populatePolicyDictionary(const CoreIPCSecTrustData::PolicyOption& options, RetainPtr<NSMutableDictionary> dict)
```
This function should create and return dictionary instead (instead of taking NSMutableDictionary)

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+                },
+                [&] (const CoreIPCDate& d) {
+                    auto date = d.toID();
+                    value = WTFMove(date);
```
`value = d.toID();`

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+                },
+                [&] (const CoreIPCString& s) {
+                    auto str = s.toID();
+                    value = WTFMove(str);
```
`value.s.toID()`

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+            ASSERT_NOT_REACHED();
+            return PolicyOptionValueShape::Invalid;
+        }
+        id element0 = a[0];
```
```
id element = [a firstObject];
if (! element) {
    ASSERT_NOT_REACHED();
    return PolicyOptionValueShape::Invalid;
}
```

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+        auto k = CoreIPCString(optionKey);
+        id optionValue = [policyOption objectForKey:optionKey];
+
+        CoreIPCSecTrust::PolicyOptionValueShape shape = CoreIPCSecTrust::detectPolicyOptionShape(optionValue);
```
Why do we need to detect policy option shape? Can't we just directly switch on `optionValue`? 
Like:
```
if ([optionValue isKindOfClass:NSString.class]) {
    CoreIPCSecTrustData::PolicyVariant value = CoreIPCString(optionValue);
    policyVector.append(std::make_pair(WTFMove(key), WTFMove(value)));
} else if ([optionValue isKindOfClass:NSNumber.class]) {
...
```

sheeparegreat replied:
I wrote it this way to reduce the level of indentation and separates the "what we have" from "what are we doing".

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,665 @@
+                ASSERT_NOT_REACHED();
+                return true;
+            }
+            id element0 = value[0];
```
`firstObject`

sheeparegreat commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,709 @@
+            secTrustData.result = CoreIPCSecTrustResult::Invalid;
+            RELEASE_LOG_ERROR(IPC, "CoreIPCSecTrust 'result' value was not set to a known constant");
+            ASSERT_NOT_REACHED();
+            break;
```
return

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,713 @@
+        return PolicyOptionValueShape::String;
+    else if ([option isKindOfClass:NSArray.class]) {
+        NSArray *a = option;
+        id element = a.firstObject;
```
Can we do just `id element = [option firstObject]`?

sheeparegreat replied:
not without a cast as option is id.  I'm OK with this

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,713 @@
+        switch (shape) {
+        case CoreIPCSecTrust::PolicyOptionValueShape::Invalid:
+            ASSERT_NOT_REACHED();
+            return true;
```
Duplicate checks

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,713 @@
+    return PolicyOptionValueShape::Invalid;
+}
+
+static bool updatePolicyVector(NSDictionary *policyOption, CoreIPCSecTrustData::PolicyOption& policyVector)
```
We could make `updatePolicyVector` return an error `String`, then on the caller side you could check the return value, and if it's not null, print it and early return.

szewai commented on Source/WebKit/Shared/cf/CoreIPCSecTrust.mm:
```diff
@@ -0,0 +1,713 @@
+        RELEASE_LOG_ERROR(IPC, "CoreIPCSecTrust 'result' value is nil or not an NSNumber and not optional");
+        return;
+    }
+ALLOW_DEPRECATED_DECLARATIONS_BEGIN
```
Which API is deprecated?

sheeparegreat committed:
Migrate SecTrust to WebKitSecureCoding Interface
https://bugs.webkit.org/show_bug.cgi?id=284795
rdar://136786582

Reviewed by Sihui Liu.

* Source/WTF/wtf/spi/cocoa/SecuritySPI.h:
* Source/WebKit/Shared/cf/CoreIPCSecTrust.h:
(WebKit::CoreIPCSecTrust::CoreIPCSecTrust):
* Source/WebKit/Shared/cf/CoreIPCSecTrust.mm: Added.
(WebKit::arrayElementsTheSameType):
(WebKit::CoreIPCSecTrust::detectPolicyOptionShape):
(WebKit::updatePolicyVector):
(WebKit::optionalArrayOfDataHelper):
(WebKit::CoreIPCSecTrust::CoreIPCSecTrust):
(WebKit::createPolicyDictionary):
(WebKit::addToDictFromOptionalDataHelper):
(WebKit::CoreIPCSecTrust::createSecTrust const):
* Source/WebKit/Shared/cf/CoreIPCSecTrust.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(secTrustRefsEqual):
(TEST(IPCSerialization, SecTrustRef)):

Canonical link: https://commits.webkit.org/289561@main
